# The DataFrame

### Objectives
After this lesson you should be able to...
+ Construct a DataFrame with the constructor
+ Know a DataFrame is composed of columns, an index and values
+ The columns are formally a column index and have the same type as the index
+ Most analysis will take place in a DataFrame
+ Read in data with **`read_csv`**
+ Change display settings in the notebook
+ Get descriptive statistics (metadata) on DataFrames after import
+ Take advantage of the index by setting it equal to a DataFrame column
+ Know why setting a DataFrame as an index is useful
+ Select a subset of columns from a DataFrame
+ Know that a DataFrame is a collection of Series
+ Create a Series by selecting a single column
+ Select rows from a DataFrame by label and position with **`loc`** and **`.iloc`**
+ Add new columns to a DataFrame
+ Know the most common column datatypes
+ Create categorical datatype with the **`cut`** function

### Prepare for this lesson by...
[ALWAYS READ THE DOCUMENTATION BEFORE A LESSON!](http://pandas.pydata.org/pandas-docs/stable/)
+ Read the [Intro to Data Structures](http://pandas.pydata.org/pandas-docs/stable/dsintro.html#dataframe) - only the DataFrame section
+ Read all of the [Options and Settings](http://pandas.pydata.org/pandas-docs/stable/options.html) section
+ Read the section on [descriptive statistics](http://pandas.pydata.org/pandas-docs/stable/basics.html#descriptive-statistics)
+ Read the first section on [Categorical Data](http://pandas.pydata.org/pandas-docs/stable/categorical.html)

### Introduction
The vast majority of the work as a data analyst will be in tabular - rows/columns form. The **DataFrame** is the primary pandas data structure and can be thought of as a collection of Series. Each row and column of the DataFrame have their own **Index**. The **Index** for the columns is simply referred to as the **column names**. Operations on DataFrames can be applied to all elements or by row or by column. The technical term **axis** refers to the horizontal and vertical components of the frame. 

The row axis is numbered 0 and the column axis is numbered 1, which is convention borrowed from NumPy where ndarrays can have limitless axes beginning with 0. The **`axis`** argument shows up in nearly all DataFrame methods, meaning you can choose to do the operation over the columns or the rows.

Just as with Series, alignment of indexes silently takes place behind the scenes for DataFrames, so care needs to be taken when operating on 2 different DataFrames at the same time.

### Manually construct a DataFrame
It is quite rare that you will actually need to construct a DataFrame manually.  Most frequently you will be reading external flat files, getting data from the web or reading from relational databases. Nonetheless, it does occur.

Just like the **`pd.Series`** is technically a constructor, **`pd.DataFrame`** constructs a DataFrame by accepting a dictionary with strings as the **keys** and lists as the **values**.

In [1]:
# Lets import our packages.
import pandas as pd
import numpy as np

In [2]:
# create dataframe from a dictionary of lists. The keys are the column names NOT the indices
df = pd.DataFrame({'name':['Ted', 'Ned', 'Jed'], 'Phone':['Samsung', 'Samsung', 'IOS'], 'Favorite Number':[99, 7, 4]})
df

Favorite Number    Phone name
0               99  Samsung  Ted
1                7  Samsung  Ned
2                4      IOS  Jed

### Closely Examine Output
It is clear that there are three columns and three rows to the DataFrame. What's not as obvious is that there exists an index for the rows with labels 0,1,2. The column names (which are also Index objects) are Favorite Number, Phone and name.

### Why did the column order get changed?
If you looked closely, the order of the columns did not match the order in which they were written in the dictionary. Dictionaries are inherently unordered, so it isn't likely that the order of the keys in the dictionary will be the same as the in the DataFrame produced from it.

It is possible to specify the order explicitly in the constructor with the **`columns`** parameter.

In [3]:
# Let's fix the column order 
df = pd.DataFrame({'name':['Ted', 'Ned', 'Jed'], 'Phone':['Samsung', 'Samsung', 'IOS'], 'Favorite Number':[99, 7, 4]},
                 columns=['name', 'Phone', 'Favorite Number'])
df

name    Phone  Favorite Number
0  Ted  Samsung               99
1  Ned  Samsung                7
2  Jed      IOS                4

### More info on the DataFrame
When a Series is outputted, the data type of the values is printed to the screen. This doesn't happen with a DataFrame. Use the **`info`** method to get more info.

In [4]:
# get more info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
name               3 non-null object
Phone              3 non-null object
Favorite Number    3 non-null int64
dtypes: int64(1), object(2)
memory usage: 152.0+ bytes


Lots of info above: the object type (DataFrame), the row index type(a RangeIndex), the column names and their types ('object' is used for strings) and whether there are any **null** values. A summary of the datatypes (2 object and 1 integer) and the memory usage.

### DataFrame by Pieces
Each DataFrame has both a column **Index** and a row **Index**. The **values** (the actual data) are represented by a 2 dimensional NumPy array. Let's see those pieces.

In [5]:
# Row index is retrieved by .index
df.index

RangeIndex(start=0, stop=3, step=1)

In [6]:
# column index is just .columns
df.columns

Index(['name', 'Phone', 'Favorite Number'], dtype='object')

In [7]:
# the values are a numpy array
df.values

array([['Ted', 'Samsung', 99],
       ['Ned', 'Samsung', 7],
       ['Jed', 'IOS', 4]], dtype=object)

### Get summary statistics for numeric columns
Basic summary statistics for numeric columns can be retrieved through the **`describe`** method. **`describe`** ignores non-numeric columns

In [8]:
df.describe()

Favorite Number
count         3.000000
mean         36.666667
std          54.003086
min           4.000000
25%           5.500000
50%           7.000000
75%          53.000000
max          99.000000

Different summary statistics for columns with characters ('objects') can be retrieved with the following

In [9]:
# get summary statistics on only the 'objects'
df.describe(include=['object'])

name    Phone
count     3        3
unique    3        2
top     Ted  Samsung
freq      1        2

### Creating a DataFrame from NumPy
An easy way to create a DataFrame is using a 2-d numpy array. Below, a 5x10 numpy array is used with both column and row indexes provided.

In [10]:
df = pd.DataFrame(np.random.rand(10,5), columns=['one', 'two', 'three', 'four', 'five'], index=list('abcdefghij'))
df

one       two     three      four      five
a  0.302376  0.808984  0.740194  0.682046  0.665099
b  0.027281  0.833645  0.805993  0.963395  0.003285
c  0.794051  0.819488  0.476746  0.980820  0.415232
d  0.147107  0.938804  0.116158  0.602068  0.538863
e  0.778638  0.399095  0.773855  0.462923  0.602969
f  0.797870  0.232582  0.614100  0.536607  0.510853
g  0.247915  0.662210  0.002795  0.288264  0.213154
h  0.507118  0.075962  0.106191  0.685351  0.357375
i  0.203346  0.745923  0.300464  0.136634  0.540480
j  0.938579  0.022483  0.585015  0.779027  0.128986

## Your data is horrible. Can we speed it up and get some real data?
Yes, there is a dataset that was downloaded from the [city of Houston](http://data.ohouston.org/dataset/city-of-houston-current-employee-roster) with employee data. Information for 2000 employees such as name, department, salary, race and others are provided. The **`read_csv`** pandas function reads csv files into a pandas DataFrame object. There are many options in the function for handling the wide variety of csvs that you will encounter in the wild.

In [11]:
# There are numerous options to read_csv but this is already a clean dataset so we don't use them now
df_coh = pd.read_csv('data/coh_employee.csv')

### Standard operations on a new dataset
When reading in a new dataset, I usually run a standard set of commands to get basic information. **`shape, head, info, describe`** 

In [12]:
# get the dimensions of the dataset
df_coh.shape

(2000, 31)

In [13]:
#lets get a small glimpse of it
df_coh.head(10)

UNIQUE_ID FIRST_NAME   LAST_NAME  POSITION_NUMBER  POSITION_JOB_CODE  \
0       9172    LILLIAN      WARDEN         30027193              306.2   
1      12311  ESPERANZA   RODRIGUEZ         30044399              901.2   
2      28693   JONATHAN  JORSCH JR.         30063393              108.0   
3       2359       JOHN        HARE         30007418              103.2   
4       5123    RUSSELL   GALBREATH         30006294              523.2   
5       3688      CRAIG    PENNAMON         30005060              108.2   
6      25295      PETER      TRUONG         30049153              778.4   
7      28212    JEFFREY     DELLING         30043726              520.3   
8       8010     ROBERT       OAKES         30051248              306.3   
9      32504     DAMIAN     LEFFLER         30019784              926.3   

                             POSITION_TITLE  BUSINESS_AREA  \
0               ASSISTANT DIRECTOR (EX LVL)           1600   
1                         LIBRARY ASSISTANT           3400   
2                            POLICE OFFICER           1000   
3                         ENGINEER/OPERATOR           1200   
4                               ELECTRICIAN           2500   
5                     SENIOR POLICE OFFICER           1000   
6                                  ENGINEER           2000   
7                                 CARPENTER           2800   
8  DEPUTY ASSISTANT DIRECTOR (EXECUTIVE LEV           2000   
9            AIRPORT OPERATIONS COORDINATOR           2800   

                       DEPARTMENT PAY_GRADE  FUND_ID      ...       \
0     Municipal Courts Department        32     1000      ...        
1                         Library        05     1000      ...        
2   Houston Police Department-HPD      PA03     1000      ...        
3   Houston Fire Department (HFD)      FD04     1000      ...        
4     General Services Department        18     2105      ...        
5   Houston Police Department-HPD      PA04     1000      ...        
6  Public Works & Engineering-PWE        26     8300      ...        
7    Houston Airport System (HAS)        14     8001      ...        
8  Public Works & Engineering-PWE        30     2301      ...        
9    Houston Airport System (HAS)        20     8001      ...        

  EMPLOYMENT_SUB_GROUP  EXEMPT  GENDER  EMPLOYMENT_STATUS CIVIL_SERVICE_TYPE  \
0        Exempt Excptn       E  Female             Active           CIVILIAN   
1     Non Exempt Postv       N  Female             Active           CIVILIAN   
2    Non-Exempt Excptn       N    Male             Active         CLASSIFIED   
3     NE Suprsn Excptn       N    Male             Active         CLASSIFIED   
4     Non Exempt Postv       N    Male             Active           CIVILIAN   
5    Non-Exempt Excptn       N    Male             Active         CLASSIFIED   
6        Exempt Excptn       E    Male             Active           CIVILIAN   
7     Non Exempt Postv       N    Male             Active           CIVILIAN   
8        Exempt Excptn       E    Male             Active           CIVILIAN   
9     Non Exempt Postv       N    Male             Active           CIVILIAN   

    HIRE_DATE    JOB_DATE   COMP_DATE                           EEOJ  \
0  2006-06-12  2012-10-13  1998-05-13  A: Officials & Administrators   
1  2000-07-19  2010-09-18  2000-07-19          E: Para-Professionals   
2  2015-02-03  2015-02-03  2014-02-03  D: Protective Service Workers   
3  1982-02-08  1991-05-25  1982-02-08  D: Protective Service Workers   
4  1989-06-19  1994-10-22  1989-06-19       G: Skilled Craft Workers   
5  1984-11-26  2005-03-26  1984-11-26  D: Protective Service Workers   
6  2012-03-26  2012-03-26  2012-03-26               B: Professionals   
7  2013-11-04  2013-11-04  2013-11-04       G: Skilled Craft Workers   
8  1993-11-15  2013-01-05  1993-11-15  A: Officials & Administrators   
9  2016-03-14  2016-03-14  2016-03-14               B: Professionals   

  SNAPSHOT_DATE  
0    2016-06-01  
1    2016-06-

### Changing display settings
If you scroll to the right, you will notice that not all 31 columns are displayed to the screen. Pandas comes with default values for a couple dozen display settings to help control output. One of these parameters is the number of columns displayed to the screen. The options can all be found under **`pd.options.display`**. The all the display settings that can be changed below.

In [14]:
dir(pd.options.display)

['chop_threshold',
 'colheader_justify',
 'column_space',
 'date_dayfirst',
 'date_yearfirst',
 'encoding',
 'expand_frame_repr',
 'float_format',
 'height',
 'large_repr',
 'latex',
 'line_width',
 'max_categories',
 'max_columns',
 'max_colwidth',
 'max_info_columns',
 'max_info_rows',
 'max_rows',
 'max_seq_items',
 'memory_usage',
 'mpl_style',
 'multi_sparse',
 'notebook_repr_html',
 'pprint_nest_depth',
 'precision',
 'show_dimensions',
 'unicode',
 'width']

In [15]:
# Lets see the current settings
pd.options.display.max_columns

20

Only 20 of the 30 columns are printed out in the notebook. We can adjust this setting easily by reassigning its value. More info on [pandas options settings](http://pandas.pydata.org/pandas-docs/stable/options.html).

In [16]:
# reassign max_columns
pd.options.display.max_columns = 40

In [17]:
# and now inspect the dataframe again
df_coh.head()

UNIQUE_ID FIRST_NAME   LAST_NAME  POSITION_NUMBER  POSITION_JOB_CODE  \
0       9172    LILLIAN      WARDEN         30027193              306.2   
1      12311  ESPERANZA   RODRIGUEZ         30044399              901.2   
2      28693   JONATHAN  JORSCH JR.         30063393              108.0   
3       2359       JOHN        HARE         30007418              103.2   
4       5123    RUSSELL   GALBREATH         30006294              523.2   

                POSITION_TITLE  BUSINESS_AREA                     DEPARTMENT  \
0  ASSISTANT DIRECTOR (EX LVL)           1600    Municipal Courts Department   
1            LIBRARY ASSISTANT           3400                        Library   
2               POLICE OFFICER           1000  Houston Police Department-HPD   
3            ENGINEER/OPERATOR           1200  Houston Fire Department (HFD)   
4                  ELECTRICIAN           2500    General Services Department   

  PAY_GRADE  FUND_ID                                 FUND_NAME  COST_CENTER  \
0        32     1000                              General Fund   1600010001   
1        05     1000                              General Fund   3400070001   
2      PA03     1000                              General Fund   1000010068   
3      FD04     1000                              General Fund   1200030001   
4        18     2105  Maintenance Renewal and Replacement Fund   2500100003   

       COST_CENTER_NAME  REPORTS_TO_POSITION        MANAGER_NAME  BASE_SALARY  \
0    MCD-Admin Services           30044747.0        NELLY SANTOS     121862.0   
1    HPL-Neigh Lib Syst           30044159.0  MIGDALIA SEPULVEDA      26125.0   
2  HPD-Vehicular Crimes           30060941.0    CATRINA WILLIAMS      45279.0   
3        HFD-Deployment                  NaN                 NaN      63166.0   
4         GSD-MRR-Maint           30008673.0         JOHN BOGNEY      56347.0   

       ORG_UNIT                      ORG_UNIT_NAME  \
0       16.1410        MCD-ADMINISTRATIVE SERVICES   
1    34.7001.27  HPL-NEIGHBORHOOD LIBRARY SERVICES   
2  10.1687.0130      HPD-VEHICULAR CRIMES DIVISION   
3       12.1210          HFD-FIRE & EMS OPERATIONS   
4       25.1416     GSD-PROPERTY MGMT MAINT & OPER   

                       ETHNICITY             RACE EMPLOYMENT_TYPE  \
0                       HISPANIC  Hispanic/Latino       Full Time   
1                       HISPANIC  Hispanic/Latino       Full Time   
2  WHITE, NOT OF HISPANIC ORIGIN            White       Full Time   
3  WHITE, NOT OF HISPANIC ORIGIN            White       Full Time   
4  WHITE, NOT OF HISPANIC ORIGIN            White       Full Time   

  EMPLOYMENT_SUB_GROUP EXEMPT  GENDER EMPLOYMENT_STATUS CIVIL_SERVICE_TYPE  \
0        Exempt Excptn      E  Female            Active           CIVILIAN   
1     Non Exempt Postv      N  Female            Active           CIVILIAN   
2    Non-Exempt Excptn      N    Male            Active         CLASSIFIED   
3     NE Suprsn Excptn      N    Male            Active         CLASSIFIED   
4     Non Exempt Postv      N    Male            Active           CIVILIAN   

    HIRE_DATE    JOB_DATE   COMP_DATE                           EEOJ  \
0  2006-06-12  2012-10-13  1998-05-13  A: Officials & Administrators   
1  2000-07-19  2010-09-18  2000-07-19          E: Para-Professionals   
2  2015-02-03  2015-02-03  2014-02-03  D: Protective Service Workers   
3  1982-02-08  1991-05-25  1982-02-08  D: Protective Service Workers   
4  1989-06-19  1994-10-22  1989-06-19       G: Skilled Craft Workers   

  SNAPSHOT_DATE  
0    2016-06-01  
1    2016-06-01  
2    2016-06-01  
3    2016-06-01  
4    2016-06-01

In [18]:
# continue by getting the 'info'
df_coh.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 31 columns):
UNIQUE_ID               2000 non-null int64
FIRST_NAME              2000 non-null object
LAST_NAME               2000 non-null object
POSITION_NUMBER         2000 non-null int64
POSITION_JOB_CODE       2000 non-null float64
POSITION_TITLE          2000 non-null object
BUSINESS_AREA           2000 non-null int64
DEPARTMENT              2000 non-null object
PAY_GRADE               2000 non-null object
FUND_ID                 2000 non-null int64
FUND_NAME               2000 non-null object
COST_CENTER             2000 non-null int64
COST_CENTER_NAME        2000 non-null object
REPORTS_TO_POSITION     1688 non-null float64
MANAGER_NAME            1678 non-null object
BASE_SALARY             1886 non-null float64
ORG_UNIT                2000 non-null object
ORG_UNIT_NAME           2000 non-null object
ETHNICITY               1998 non-null object
RACE                    1965 non-null ob

There are 3 columns with float values, 5 with integer and 23 with strings. There are **missing** values in this dataset. Since there are 2000 rows, any column that has less than 2000 **non-null** values must have some missing values. For instance the **REPORTS_TO_POSITION** column has 1688 non-null values which means 312  are missing.

In [19]:
# get summary statistics. describe defaults to only numeric columns
# the only meaningful column here is BASE_SALARY
df_coh.describe()

UNIQUE_ID  POSITION_NUMBER  POSITION_JOB_CODE  BUSINESS_AREA  \
count   2000.000000     2.000000e+03        2000.000000    2000.000000   
mean   16247.652500     3.003217e+07         372.678600    2135.350000   
std     9509.483839     2.182383e+04         283.143263    1660.687906   
min        5.000000     3.000001e+07          47.000000    1000.000000   
25%     7981.000000     3.001107e+07         108.000000    1000.000000   
50%    15316.000000     3.002944e+07         303.500000    1200.000000   
75%    24480.500000     3.005436e+07         568.150000    2800.000000   
max    32967.000000     3.006737e+07         978.200000    9000.000000   

           FUND_ID   COST_CENTER  REPORTS_TO_POSITION    BASE_SALARY  
count  2000.000000  2.000000e+03         1.688000e+03    1886.000000  
mean   2401.221500  2.135389e+09         3.003167e+07   55767.931601  
std    2607.231936  1.660714e+09         2.337296e+04   21693.706679  
min    1000.000000  1.000010e+09         3.000004e+07   24960.000000  
25%    1000.000000  1.000010e+09         3.000661e+07   40170.000000  
50%    1000.000000  1.200050e+09         3.003395e+07   54461.000000  
75%    2301.000000  2.800020e+09         3.005502e+07   66614.000000  
max    9000.000000  9.000160e+09         3.006737e+07  275000.000000

In [20]:
# describe just the string columns
# these are much more interesting
df_coh.describe(include=['object'])

FIRST_NAME LAST_NAME         POSITION_TITLE  \
count        2000      2000                   2000   
unique        961      1371                    330   
top       MICHAEL   JOHNSON  SENIOR POLICE OFFICER   
freq           44        30                    220   

                           DEPARTMENT PAY_GRADE     FUND_NAME  \
count                            2000      2000          2000   
unique                             24        80            39   
top     Houston Police Department-HPD      PA03  General Fund   
freq                              638       184          1304   

       COST_CENTER_NAME     MANAGER_NAME ORG_UNIT              ORG_UNIT_NAME  \
count              2000             1678     2000                       2000   
unique              448             1201     1181                        478   
top      HFD-Deployment  HERMAN GONZALES  12.1210  HFD-FIRE & EMS OPERATIONS   
freq                297               56      288                        289   

                            ETHNICITY                       RACE  \
count                            1998                       1965   
unique                              5                          6   
top     BLACK, NOT OF HISPANIC ORIGIN  Black or African American   
freq                              715                        700   

       EMPLOYMENT_TYPE EMPLOYMENT_SUB_GROUP EXEMPT GENDER EMPLOYMENT_STATUS  \
count             2000                 2000   1999   2000              2000   
unique               5                   17      2      2                 2   
top          Full Time     Non Exempt Postv      N   Male            Active   
freq              1954                  792   1681   1397              1991   

       CIVIL_SERVICE_TYPE   HIRE_DATE    JOB_DATE   COMP_DATE  \
count                2000        2000        1997        2000   
unique                  3         999         947        1084   
top              CIVILIAN  2016-03-28  2002-01-05  2012-04-16   
freq                 1100          11          34          10   

                                 EEOJ SNAPSHOT_DATE  
count                            1999          2000  
unique                              8             1  
top     D: Protective Service Workers    2016-06-01  
freq                              728          2000

### Transpose a DataFrame
Occasionally its useful to transpose a DataFrame - swap the columns and the rows. In this case, the output will be much more readable. Use the **`.T`** attribute.

In [21]:
df_coh.describe(include=['object']).T

count unique                            top  freq
FIRST_NAME            2000    961                        MICHAEL    44
LAST_NAME             2000   1371                        JOHNSON    30
POSITION_TITLE        2000    330          SENIOR POLICE OFFICER   220
DEPARTMENT            2000     24  Houston Police Department-HPD   638
PAY_GRADE             2000     80                           PA03   184
FUND_NAME             2000     39                   General Fund  1304
COST_CENTER_NAME      2000    448                 HFD-Deployment   297
MANAGER_NAME          1678   1201                HERMAN GONZALES    56
ORG_UNIT              2000   1181                        12.1210   288
ORG_UNIT_NAME         2000    478      HFD-FIRE & EMS OPERATIONS   289
ETHNICITY             1998      5  BLACK, NOT OF HISPANIC ORIGIN   715
RACE                  1965      6      Black or African American   700
EMPLOYMENT_TYPE       2000      5                      Full Time  1954
EMPLOYMENT_SUB_GROUP  2000     17               Non Exempt Postv   792
EXEMPT                1999      2                              N  1681
GENDER                2000      2                           Male  1397
EMPLOYMENT_STATUS     2000      2                         Active  1991
CIVIL_SERVICE_TYPE    2000      3                       CIVILIAN  1100
HIRE_DATE             2000    999                     2016-03-28    11
JOB_DATE              1997    947                     2002-01-05    34
COMP_DATE             2000   1084                     2012-04-16    10
EEOJ                  1999      8  D: Protective Service Workers   728
SNAPSHOT_DATE         2000      1                     2016-06-01  2000

### Always know your Index
Its important to be aware of what your (row) index is in a DataFrame. The column index is usually obvious as it's just the column names.

In [22]:
# when the read_csv is not given an index, the default is make the index integers starting from 0
df_coh.index

RangeIndex(start=0, stop=2000, step=1)

In [23]:
# See the values of a RangeIndex
df_coh.index.values

array([   0,    1,    2, ..., 1997, 1998, 1999], dtype=int64)

In [24]:
# the column index is also just a numpy array
df_coh.columns.values

array(['UNIQUE_ID', 'FIRST_NAME', 'LAST_NAME', 'POSITION_NUMBER',
       'POSITION_JOB_CODE', 'POSITION_TITLE', 'BUSINESS_AREA',
       'DEPARTMENT', 'PAY_GRADE', 'FUND_ID', 'FUND_NAME', 'COST_CENTER',
       'COST_CENTER_NAME', 'REPORTS_TO_POSITION', 'MANAGER_NAME',
       'BASE_SALARY', 'ORG_UNIT', 'ORG_UNIT_NAME', 'ETHNICITY', 'RACE',
       'EMPLOYMENT_TYPE', 'EMPLOYMENT_SUB_GROUP', 'EXEMPT', 'GENDER',
       'EMPLOYMENT_STATUS', 'CIVIL_SERVICE_TYPE', 'HIRE_DATE', 'JOB_DATE',
       'COMP_DATE', 'EEOJ', 'SNAPSHOT_DATE'], dtype=object)

## The [ ] is completely different for DataFrames than for Series
The bracket's primary use is to retrieve a column(s) from a DataFrame. Simply write the name of the column into the brackets and a Series will be returned. This behavior for the [ ] is completely different for Series where it is used to retrieve value(s) by the index.

In [25]:
# Access just the FIRST_NAME column
# shorten output with head
# returns a Series

df_coh['FIRST_NAME'].head(10)

0      LILLIAN
1    ESPERANZA
2     JONATHAN
3         JOHN
4      RUSSELL
5        CRAIG
6        PETER
7      JEFFREY
8       ROBERT
9       DAMIAN
Name: FIRST_NAME, dtype: object

### Use a list to access multiple columns

In [26]:
# get three columns - put names in a list
# Returns a DataFrame

df_coh[['FIRST_NAME', 'GENDER', 'RACE']].head()

FIRST_NAME  GENDER             RACE
0    LILLIAN  Female  Hispanic/Latino
1  ESPERANZA  Female  Hispanic/Latino
2   JONATHAN    Male            White
3       JOHN    Male            White
4    RUSSELL    Male            White

In [27]:
# If you want to return a single column as a dataframe use a list with one column name
df_coh[['FIRST_NAME']].head(10)

FIRST_NAME
0    LILLIAN
1  ESPERANZA
2   JONATHAN
3       JOHN
4    RUSSELL
5      CRAIG
6      PETER
7    JEFFREY
8     ROBERT
9     DAMIAN

### You have a Series when you access a single column
It's important to know that when you use the **`[ ]`** operator for a DataFrame and pass a single column name into it, you are back to having a Series. All the Series functions will work as before.

In [28]:
# use the same Series functions as before

# use max series method
df_coh['BASE_SALARY'].max()

275000.0

In [29]:
# sum total salary
df_coh['BASE_SALARY'].sum()

105178319.0

In [30]:
# find the standard deviation in salaries
df_coh['BASE_SALARY'].std()

21693.706679449504

In [31]:
# use boolean selection to select salaries > 100,000
salaries = df_coh['BASE_SALARY']
criteria = salaries > 100000

salaries[criteria].head(10)

0      121862.0
8      107962.0
11     180416.0
43     165216.0
66     100791.0
169    120916.0
178    210588.0
186    110881.0
217    102019.0
237    130416.0
Name: BASE_SALARY, dtype: float64

In [32]:
# you can also use the criteria to filter the entire dataframe
# All columns from those with a salary > 100,000

df_coh[criteria].head()

UNIQUE_ID FIRST_NAME  LAST_NAME  POSITION_NUMBER  POSITION_JOB_CODE  \
0        9172    LILLIAN     WARDEN         30027193              306.2   
8        8010     ROBERT      OAKES         30051248              306.3   
11       4560     LUTHER    HARRELL         30002735              724.2   
43      27959    MICHAEL   GONZALEZ         30058271              656.6   
66      32229      SALLY  ABOUASSAF         30058430              706.5   

                              POSITION_TITLE  BUSINESS_AREA  \
0                ASSISTANT DIRECTOR (EX LVL)           1600   
8   DEPUTY ASSISTANT DIRECTOR (EXECUTIVE LEV           2000   
11                        CHIEF PHYSICIAN,MD           3800   
43          ASSOCIATE EMS PHYSICIAN DIRECTOR           1200   
66                 PUBLIC HEALTH DENTIST,DDS           3800   

                        DEPARTMENT PAY_GRADE  FUND_ID  \
0      Municipal Courts Department        32     1000   
8   Public Works & Engineering-PWE        30     2301   
11         Health & Human Services        35     1000   
43   Houston Fire Department (HFD)        35     2010   
66         Health & Human Services        29     2010   

                           FUND_NAME  COST_CENTER      COST_CENTER_NAME  \
0                       General Fund   1600010001    MCD-Admin Services   
8           Building Inspection Fund   2000060004  PWE-BldgOfficialOffi   
11                      General Fund   3800050025   HHS-CHS AreaOprSppt   
43  Essential Public Health Services   1200030004  HFD-Medical Dir/Q.A.   
66  Essential Public Health Services   3800070010  HHS-GeriatricOralHlt   

    REPORTS_TO_POSITION    MANAGER_NAME  BASE_SALARY    ORG_UNIT  \
0            30044747.0    NELLY SANTOS     121862.0     16.1410   
8            30043470.0      EARL GREER     107962.0  20.2123.01   
11           30048040.0     RISHA JONES     180416.0     38.5038   
43           30009179.0    DAVID PERSSE     165216.0  12.1225.03   
66           30021232.0  DEBORAH GARCIA     100791.0     38.5023   

                      ORG_UNIT_NAME                      ETHNICITY  \
0       MCD-ADMINISTRATIVE SERVICES                       HISPANIC   
8   PWE-OFFICE OF THE BLDG OFFICIAL  WHITE, NOT OF HISPANIC ORIGIN   
11             HHS-FAMILY HLTH ADMN  BLACK, NOT OF HISPANIC ORIGIN   
43     HFD-MEDICAL DIRECTION & Q.A.                       HISPANIC   
66                  HHS-ORAL HEALTH  WHITE, NOT OF HISPANIC ORIGIN   

                         RACE EMPLOYMENT_TYPE EMPLOYMENT_SUB_GROUP EXEMPT  \
0             Hispanic/Latino       Full Time        Exempt Excptn      E   
8                       White       Full Time        Exempt Excptn      E   
11  Black or African American       Full Time        Exempt Excptn      E   
43            Hispanic/Latino       Full Time        Exempt Excptn      E   
66                      White       Full Time        Exempt Excptn      E   

    GENDER EMPLOYMENT_STATUS CIVIL_SERVICE_TYPE   HIRE_DATE    JOB_DATE  \
0   Female            Active           CIVILIAN  2006-06-12  2012-10-13   
8     Male            Active           CIVILIAN  1993-11-15  2013-01-05   
11    Male            Active           CIVILIAN  1987-05-22  1999-08-28   
43    Male            Active           CIVILIAN  2013-08-31  2013-08-31   
66  Female            Active           CIVILIAN  2015-12-28  2015-12-28   

     COMP_DATE                           EEOJ SNAPSHOT_DATE  
0   1998-05-13  A: Officials & Administrators    2016-06-01  
8   1993-11-15  A: Officials & Administrators    2016-06-01  
11  1987-05-22  A: Officials & Administrators    2016-06-01  
43  2013-08-31  A: Officials & Administrators    2016-06-01  
66  2015-12-28               B: Professionals    2016-06-01

### More interesting boolean logic with DataFrames
Many more interesting questions can now be asked and solved with boolean operations in the data in a very similar manner to how they were in the Series notebook.

In [33]:
# find people with make more 100,000 and are female
criteria = (df_coh['BASE_SALARY'] > 100000) & (df_coh['GENDER'] == 'Female')

df_coh[criteria].head()

UNIQUE_ID FIRST_NAME  LAST_NAME  POSITION_NUMBER  POSITION_JOB_CODE  \
0         9172    LILLIAN     WARDEN         30027193              306.2   
66       32229      SALLY  ABOUASSAF         30058430              706.5   
237      10297      MARIA     IRSHAD         30057398              306.2   
366      29194    ELEANOR     HOLMAN         30060598              306.3   
522      24914   KERTECIA       MOND         30049334              306.3   

                               POSITION_TITLE  BUSINESS_AREA  \
0                 ASSISTANT DIRECTOR (EX LVL)           1600   
66                  PUBLIC HEALTH DENTIST,DDS           3800   
237      ASSISTANT DIRECTOR (EXECUTIVE LEVEL)           6500   
366  DEPUTY ASSISTANT DIRECTOR (EXECUTIVE LEV           5000   
522        DEPUTY ASSISTANT DIRECTOR (EX LVL)           2800   

                       DEPARTMENT PAY_GRADE  FUND_ID  \
0     Municipal Courts Department        32     1000   
66        Health & Human Services        29     2010   
237    Admn. & Regulatory Affairs        32     8700   
366                Mayor's Office        30     2429   
522  Houston Airport System (HAS)        30     8001   

                            FUND_NAME  COST_CENTER      COST_CENTER_NAME  \
0                        General Fund   1600010001    MCD-Admin Services   
66   Essential Public Health Services   3800070010  HHS-GeriatricOralHlt   
237  Parking Mangement Operating Fund   6500090001  ARA-Parking Cust Ser   
366         Houston Civic Events Fund   5000030001  MYR-Hou Civic Events   
522                  HAS-Revenue Fund   2800010003   HAS-DO Intern Audit   

     REPORTS_TO_POSITION     MANAGER_NAME  BASE_SALARY      ORG_UNIT  \
0             30044747.0     NELLY SANTOS     121862.0       16.1410   
66            30021232.0   DEBORAH GARCIA     100791.0       38.5023   
237           30030918.0   ERNESTINA PAEZ     130416.0       65.1700   
366           30062588.0  SUSAN CHRISTIAN     110000.0    50.1430.05   
522           30047588.0  BALRAM BHEODARI     110686.0  28.1113.1000   

                   ORG_UNIT_NAME                      ETHNICITY  \
0    MCD-ADMINISTRATIVE SERVICES                       HISPANIC   
66               HHS-ORAL HEALTH  WHITE, NOT OF HISPANIC ORIGIN   
237       ARA-PARKING MANAGEMENT      ASIAN OR PACIFIC ISLANDER   
366           MYR-SPECIAL EVENTS  WHITE, NOT OF HISPANIC ORIGIN   
522        HAS-DO INTERNAL AUDIT  BLACK, NOT OF HISPANIC ORIGIN   

                          RACE EMPLOYMENT_TYPE EMPLOYMENT_SUB_GROUP EXEMPT  \
0              Hispanic/Latino       Full Time        Exempt Excptn      E   
66                       White       Full Time        Exempt Excptn      E   
237     Asian/Pacific Islander       Full Time        Exempt Excptn      E   
366                      White       Full Time        Exempt Excptn      E   
522  Black or African American       Full Time        Exempt Excptn      E   

     GENDER EMPLOYMENT_STATUS CIVIL_SERVICE_TYPE   HIRE_DATE    JOB_DATE  \
0    Female            Active           CIVILIAN  2006-06-12  2012-10-13   
66   Female            Active           CIVILIAN  2015-12-28  2015-12-28   
237  Female            Active           CIVILIAN  2002-05-24  2013-07-20   
366  Female            Active           CIVILIAN  2014-05-13  2014-05-13   
522  Female            Active           CIVILIAN  2011-11-07  2011-11-07   

      COMP_DATE                           EEOJ SNAPSHOT_DATE  
0    1998-05-13  A: Officials & Administrators    2016-06-01  
66   2015-12-28               B: Professionals    2016-06-01  
237  2002-05-24  A: Officials & Administrators    2016-06-01  
366  2014-05-13  A: Officials & Administrators    2016-06-01  
522  2011-11-07  A: Officials & Administrators    2016-06-01

In [34]:
df_coh['RACE'].value_counts()

Black or African American            700
White                                665
Hispanic/Latino                      480
Asian/Pacific Islander               107
American Indian or Alaskan Native     11
Others                                 2
Name: RACE, dtype: int64

In [35]:
# your code here
white_female = 
black_male = 
over_100k = 

criteria = 

SyntaxError: invalid syntax (<ipython-input-35-72293e33b01b>, line 2)

### Problem 6
<span  style="color:green; font-size:16px">What is the third most common department?</span>

In [ ]:
# your code here

### Do men or woman make more money?
Boolean indexing can help answer this question. We first create two new DataFrames that are filtered for each gender and then find the mean of the salary. Men make about $5,000 more.

In [ ]:
men = df_coh[df_coh['GENDER'] == 'Male']
women = df_coh[df_coh['GENDER'] == 'Female']

men['BASE_SALARY'].mean(), women['BASE_SALARY'].mean()

### All unique values of a column
Frequently it's nice to know all the possible values contained in a column. I usually use the **`value_counts`** method which counts the frequencies but also available is the **`unique`** method.

In [ ]:
# Get all unique values of RACE
df_coh['RACE'].unique()

In [ ]:
# but I prefer value_counts which returns frequencies as well
df_coh['RACE'].value_counts()

### Problem 7
<span  style="color:green; font-size:16px">Who makes more money, 'Black or African American' Females or White Males? Use RACE and GENDER columns.</span>

In [ ]:
# your code here

### Problem 8
<span  style="color:green; font-size:16px">Who makes more money, 'POLICE OFFICER', 'ENGINEER/OPERATOR' or 'ELECTRICIAN'? Use the POSITION_TITLE column.</span>

In [ ]:
# your code here

### Is there a better way to answer this last couple questions?
Yes! The **groupby** method allows for grouping of these categorical variables and will be explained in greater detail in a future lesson.

### Selecting a single employee
Many datasets have a column with an integer that uniquely identifies each row.  In database speak this column is called the table's **primary key**. The primary key allows for easy and direct access to each employee.

In the **df_coh** table it appears that the first column, **UNIQUE_ID** is the primary key. Most good datasets will have a **data dictionary** that describes each column of the table so you won't have to take a guess as to what the primary key is. The data dictionary (a.k.a metadata) for the current dataset can be [found here.](http://data.ohouston.org/dataset/city-of-houston-current-employee-roster/resource/98448c04-e76f-4fa0-8916-12786e6e5883)

### Ensuring uniqueness
The most important aspect of a primary key is its uniqueness.

In [ ]:
# ensure that unique_id is unique
# use nunique Series method

df_coh['UNIQUE_ID'].nunique()

In [ ]:
# 2000 is the total number of rows so unique_id is indeed unique
len(df_coh)

### Using boolean indexing to select an employee
Now that it is clear that **UNIQUE_ID** is indeed unique, let's practice selecting Employee **28693** using boolean indexing.

### Problem 9
<span  style="color:green; font-size:16px">Select employee 28693.</span>

In [ ]:
# your code here

### Taking advantage of the index
Selecting an employee with **`df_coh[df_coh['UNIQUE_ID'] == 28693]`** is a bit cumbersome. A more elegant way involves replacing the current index with **UNIQUE_ID**. This is done with the **set_index** method and allows for more straightforward employee selections.

In [ ]:
df_coh = pd.read_csv('data/coh_employee.csv')

In [ ]:
# use set_index
# must assign result to save it
df_coh = df_coh.set_index('UNIQUE_ID')

In [ ]:
# inspect frame
df_coh.head()

### Inspecting the new DataFrame output
The index is now meaningful. The index, which previously was just a range beginning at 0, is now the employee ID. The name **UNIQUE_ID** is now the **name** of the index. The values for the index are still **bold** and a reminder that these values are part of the index and not a column. 

You should notice the blank first 'row' in the new DataFrame above. This is actually not a row at all but extra space that is inserted because the index now has a name.

In [ ]:
# the index object has a name attribute
df_coh.index.name

In [ ]:
# if that first empty row bothers you, you can delete the name of the index
del df_coh.index.name

In [ ]:
# index still has the same meaning but does not have a name.
# empty row space is now gone

df_coh.head()

### Setting the index on read
When first reading the dataset using **`read_csv`**, use the argument **`index_col`** to pass the position of the column to use as the index.

In [ ]:
df_coh = pd.read_csv('data/coh_employee.csv', index_col=0)
df_coh.head()

### Selecting a subset of rows
Selecting columns was just covered above and is done using the **`[ ]`** operator. Rows can be selected using **`.iloc`** and **`.loc`** as was done with the Series.

In [ ]:
# get the first 3 rows by integer location
df_coh.iloc[:3]

In [ ]:
# get row 155
# returns a Series

df_coh.iloc[155]

### Problem 10
<span  style="color:green; font-size:16px">Select rows 7, 77 and 777 from **`df_coh`**.</span>

In [ ]:
# your code here

In [ ]:
# Select using the index label
df_coh.loc[28693]

In [ ]:
# the above returns a series
# if you want a dataframe, wrap lables in a list

df_coh.loc[[28693]]

Notice how selecting an employee with the **`UNIQUE_ID`** in the index with **`.loc`** is easier than with boolean indexing.

In [ ]:
# it is actually possible to slice from one index label to another
# but you must know the ordering

df_coh.loc[9172:28693]

In [ ]:
# see the same as above but reversed. Nothing is selected
df_coh.loc[28693:9172]

### Problem 11
<span  style="color:green; font-size:16px">Select employees with IDs 3105, 24767 and 31578.</span>

In [ ]:
# your code here

### Selecting a subset of rows and columns together
We now have looked at independently selecting subsets of columns and rows. We will now select subsets of both columns and rows together with **`.iloc`** and **`.loc`**.  Withing the brackets, a **comma** is used to separate the rows references (which come first) from the column references.

In [ ]:
# using integer location select rows 100:105 and the first 5 columns
df_coh.iloc[100:105, :5]

### Problem 12
<span  style="color:green; font-size:16px">Brackets are used to select disjoint rows or columns. Select rows 10, 100 and 500 along with columns 3, 7 and 20.</span>

In [ ]:
# your code here

### Problem 13
<span  style="color:green; font-size:16px">Use label based indexing to select employee IDs 12214 and 8317 along with columns HIRE_DATE and GENDER.</span>

In [ ]:
# your code here

### Problem 14: Advanced
<span  style="color:green; font-size:16px">Use label based indexing to select employee IDs 12214 and 8317 along with columns all columns except FIRST_NAME and POSITION_NUMBER. [Use this stackoverflow answer for guidance.](http://stackoverflow.com/a/37441204/3707607)</span>

In [ ]:
# your code here

### Some confusion with brackets vs .iloc vs .loc
It was written earlier that only column names go inside of brackets for DataFrames. This isn't entirely true. It is possible to pass a slice to the brackets to return rows. I would highly recommend against doing this and use brackets only for column names, **.loc** for index labeled based selection and **.iloc** for integer position selection.

In [ ]:
# select the first 5 rows with brackets - bad idea!
df_coh[:5]

### Using **`.ix`** to mix both integer and label indexing (THIS IS DEPRECATED: DO NOT USE)
The **`.ix`** operator provides some flexibility if a case ever arises where selection needs to occur with both label and integer. I tend to avoid this operator as its still ambigious whether you are using integer or location selection when the index is integers. [Check the docs for more info](http://pandas.pydata.org/pandas-docs/stable/indexing.html#different-choices-for-indexing)

### Selecting single elements with `.at` and `.iat`
In the rare case that you would like to select exactly one cell of data, you can use **`.at`** for label based selection and **`.iat`** for integer based selection. They work analogously to **`.loc`** and **`.iloc`** and don't provide any extra functionality just are faster. Technically, a single element is called a **scalar** value.

They don't provide any extra functionality but are faster than their counterparts.

In [ ]:
# scalar selection
df_coh.at[5123, 'GENDER']

In [ ]:
# scalar selection with .loc
df_coh.loc[5123, 'GENDER']

### Problem 15
<span  style="color:green; font-size:16px">Use **`.iat`** correctly and explain what happened.</span>

In [ ]:
# your code here

### Problem 16
<span  style="color:green; font-size:16px">Use the **`timeit`** magic command to see the speed difference between **`.at`** and **`.loc`** for the same selection. How much faster is **`.at`**?</span>

In [ ]:
# your code here

### Adding new columns to the DataFrame
A very common task during analysis is to add new columns to the working DataFrame. Usually, some operation is performed using the existing columns with the outcome added as a new column.

In [ ]:
# add a constant value as a new column
# the new column is always last

df_coh['NEW_CONSTANT_COLUMN'] = 5

In [ ]:
#inspect dataframe
# look at first three rows and last 5 columns
# sure enough, there is our new constant column

df_coh.iloc[:3, -5:]

In [ ]:
# making a new column: FULL NAME
# column of object type can be concatenated together just like strings

df_coh['FULL_NAME'] = df_coh['FIRST_NAME'] + ' ' + df_coh['LAST_NAME']

In [ ]:
# inspect the name columns to see if new column is as intended
df_coh[['FIRST_NAME', 'LAST_NAME', 'FULL_NAME']].head()

### Adding a random numeric column
Let's say the city of Houston wants to give everyone a random bonus ranging anywhere between 0 and 10% of their current base salary. To begin, we will make a column called **RANDOM_BONUS** and assign each employee a random number between 0 and .1.

In [ ]:
# create a numpy array the same length as the DataFrame
# with random values from 0 and .1

df_coh['RANDOM_BONUS'] = np.random.rand(len(df_coh)) * .1

In [ ]:
# inspect last columns

df_coh.iloc[:3, -5:]

### Data Types
Every column of the DataFrame has a data type. This was seen at the top of the notebook with the **`info`** method. To get a more clean view of the data types use the **dtypes** attribute. The most common data types are **int64, float64, bool, datetime64, timedelta64 and object**. These are [all numpy data types](https://docs.scipy.org/doc/numpy/user/basics.types.html). 

The **object** is a catch-all for any strings, complex Python data types like lists, dictionaries, etc... or any mix of strings and numbers.

The **datetime64** data type holds specific dates in time such as Jan 24, 2017 5:30 p.m. **timedelta64** holds only time values such as 6 days, 4 hours, 14 minutes and x nanoseconds. Both **datetime64** and **timedelta64** have nanosecond precision. These time data types will be discussed in great detail in later notebooks.

In [ ]:
# only look at the data types
df_coh.dtypes

### Changing Data Types
Occasionally, some columns will not be the desired type and need to be changed to a different type. A common situation occurs when dealing with dates that are read in as strings and defaulted to the **`object`** data type. The Series method **astype** will attempt to force a column to a different type.

The argument passed to **astype** must be the name of the new data type as either a string or a NumPy object. The **HIRE_DATE** column is currently a string and gets converted to a **datetime64** below along with other date columns.

In [ ]:
# using a string
df_coh['HIRE_DATE'] = df_coh['HIRE_DATE'].astype('datetime64')

# using the numpy object
df_coh['COMP_DATE'] = df_coh['COMP_DATE'].astype(np.datetime64)

# yet another way using a function with flexibility to do more things
df_coh['SNAPSHOT_DATE'] = pd.to_datetime(df_coh['SNAPSHOT_DATE'])

# Case Study: Do people with more experience make more money?
To answer this question, the number of years of experience needs to be calculated from the column **HIRE_DATE**. **datetime64** columns can be subtracted from one another. The current date will be needed and is retrieved from the **to_datetime** pandas function.

In [ ]:
# get today's date
today = pd.to_datetime('today')

today

In [ ]:
# subtract the hire date from today to get the number of days of experience
experience = today - df_coh['HIRE_DATE']

# print out head of series
experience.head()

### Converting to years
Notice that the data type is now **timedelta64** which just reprsents an amount of time in days. To convert this to years an esoteric command must be run. [See here for more detail](http://pandas.pydata.org/pandas-docs/stable/timedeltas.html#frequency-conversion)

In [ ]:
# convert to years
years_experience = experience / np.timedelta64(1, 'Y')

# inspect and check that it makes sense
years_experience.head()

In [ ]:
# Make a new column
df_coh['YEARS_EXPERIENCE'] = years_experience

### Creating categories for years of experience
It's possible to divide numerical columns into different categories based on their value. The pandas **cut** function accepts a Series or an array and a list of the edges of the **bins**. Each category can be given a **label** as well. A series is returned that is of **categorical** type - unique to Pandas. [More on categorical data](http://pandas.pydata.org/pandas-docs/stable/categorical.html)

In [ ]:
# create Series of categorical data
exp_categories = pd.cut(years_experience, bins=[0, 5, 15, 100], labels=['Novice', 'Experienced', 'Senior'])

In [ ]:
# inspect Seriers
exp_categories.head(10)

In [ ]:
# get some summary statistics
exp_categories.value_counts()

In [ ]:
# Create new column
df_coh['EXPERIENCE_LEVEL'] = exp_categories

# Section end: You should know...
* a DataFrame has an index for both the columns and the rows
* import data with **`pd.read_csv`**
* Use **`shape, head, info, describe`** as standard operations to inspect a newly imported DataFrame
* Select a single column (a Series) with brackets
* Select multiple columns with a list of column names in brackets
* Be explicit when selecting rows with **`iloc`** and **`loc`**
* Select rows and columns together with **`iloc`** and **`loc`** by separating row and column references with a **comma**
* boolean indexing with multiple column comparisons
* Use **`set_index`** to move a column into the index.
* Primary keys are good choices for the index
* Know all the most common numpy data types **`int64, float64, bool, datetime64, timedelta64 and object`**
* change data type with **`astype`** method
* Create new columns
* use the **`cut`** function to create categorical variables

# Problems
If you have not run the code from the beginning of the notebook then run the code in the block below before beginning the problems. Most of the problems will be about the city of Houston employee dataset.

In [ ]:
# run this code block first
df_coh = pd.read_csv('data/coh_employee.csv', index_col=0)
df_coh['RANDOM_BONUS'] = np.random.rand(len(df_coh)) * .1
df_coh['HIRE_DATE'] = df_coh['HIRE_DATE'].astype(np.datetime64)
today = pd.to_datetime('today')
experience = today - df_coh['HIRE_DATE']
years_experience = experience / np.timedelta64(1, 'Y')
df_coh['YEARS_EXPERIENCE'] = years_experience
exp_categories = pd.cut(years_experience, bins=[0, 5, 15, 100], labels=['Novice', 'Experienced', 'Senior'])
df_coh['EXPERIENCE_LEVEL'] = exp_categories

### Problem 1
<span  style="color:green; font-size:16px">What object is returned from the above code **`df_coh.describe(include=['object']).T`**?</span>

In [ ]:
type(df_coh.describe(include=['object']).T)

### Problem 2
<span  style="color:green; font-size:16px">What object are the values of an **`Index`**?</span>

In [ ]:
type(df_coh.index)

### Problem 3
<span  style="color:green; font-size:16px">Retrieve the LAST_NAME column and assign it to a variable. Count the number of employees with a last name of 'JOHNSON'. Do the problem again in one line of code.</span>

In [ ]:
lastnames = df_coh['LAST_NAME']
lastnames.head()
johnsons = (df_coh['LAST_NAME']=='JOHNSON').sum()
johnsons

### Problem 4
<span  style="color:green; font-size:16px">Count all the occurrences of each value in the column RACE. Remember there is single Series method that does this.</span>

In [ ]:
df_coh['RACE'].value_counts()

### Problem 5
<span  style="color:green; font-size:16px">How many employees are White Females or 'Black or African American' Males and make more than 100,000? Create variables for each of three categories below before putting them all together in one criteria.</span>

In [ ]:
# your code here
white_female = (df_coh['GENDER'] == 'Female') & (df_coh.RACE == 'White')
black_male = (df_coh.GENDER == 'Male') & (df_coh.RACE == 'Black or African American')
over_100k = df_coh.BASE_SALARY > 100000

criteria = (white_female | black_male) & over_100k
criteria.sum()




### Problem 6
<span  style="color:green; font-size:16px">What is the third most common department?</span>

In [ ]:
df_coh.DEPARTMENT.value_counts().index[2]


### Problem 7
<span  style="color:green; font-size:16px">Who makes more money, 'Black or African American' Females or White Males? Use RACE and GENDER columns.</span>

In [ ]:
df_coh['BASE_SALARY'][(df_coh['RACE']=='Black or African American') & (df_coh['GENDER'] == 'Female')].mean(),\
df_coh['BASE_SALARY'][(df_coh['RACE']=='White')& (df_coh['GENDER'] == 'Male')].mean()

### Problem 8
<span  style="color:green; font-size:16px">Who makes more money, 'POLICE OFFICER', 'ENGINEER/OPERATOR' or 'ELECTRICIAN'? Use the POSITION_TITLE column.</span>

In [ ]:
(df_coh['BASE_SALARY'][df_coh['POSITION_TITLE'] == 'POLICE OFFICER']).mean(),\
(df_coh['BASE_SALARY'][df_coh['POSITION_TITLE'] == 'ENGINEER/OPERATOR']).mean(),\
(df_coh['BASE_SALARY'][df_coh['POSITION_TITLE'] == 'ELECTRICIAN']).mean()

### Problem 9
<span  style="color:green; font-size:16px">Select employee 28693.</span>

In [ ]:
df_coh[df_coh.UNIQUE_ID == 28693]

### Problem 10
<span  style="color:green; font-size:16px">Select rows 7, 77 and 777 from **`df_coh`**.</span>

In [ ]:
df_coh.iloc[[7,77,777]]

### Problem 11
<span  style="color:green; font-size:16px">Select employees with IDs 3105, 24767 and 31578.</span>

In [ ]:
df_coh[df_coh.UNIQUE_ID.isin([3105, 24767, 31578])]

### Problem 12
<span  style="color:green; font-size:16px">Brackets are used to select disjoint rows or columns. Select rows 10, 100 and 500 along with columns 3, 7 and 20.</span>

In [ ]:
df_coh.iloc[[10,100,500],[3,7,20]]

### Problem 13
<span  style="color:green; font-size:16px">Use label based indexing to select employee IDs 12214 and 8317 along with columns HIRE_DATE and GENDER.</span>

In [ ]:
df_coh.loc[df_coh.UNIQUE_ID.isin([12214, 8317]),['HIRE_DATE','GENDER']]

### Problem 14: Advanced
<span  style="color:green; font-size:16px">Use label based indexing to select employee IDs 12214 and 8317 along with columns all columns except FIRST_NAME and POSITION_NUMBER. [Use this stackoverflow answer for guidance.](http://stackoverflow.com/a/37441204/3707607)</span>

In [ ]:
df_coh.loc[df_coh.UNIQUE_ID.isin([12214, 8317]),df_coh.columns.difference(['FIRST_NAME', 'POSITION_NUMBER'])]


### Problem 15
<span  style="color:green; font-size:16px">Use **`.iat`** correctly and explain what happened.</span>

In [ ]:
df_coh = df_coh.set_index('UNIQUE_ID')
# selects the 11th row and 6th column value
df_coh.iat[10, 5]

### Problem 16
<span  style="color:green; font-size:16px">Use the **`timeit`** magic command to see the speed difference between **`.at`** and **`.loc`** for the same selection. How much faster is **`.at`**?</span>

In [ ]:
%timeit df_coh.loc[5123, 'GENDER']

In [ ]:
%timeit df_coh.at[5123, 'GENDER']

### Problem 17
<span  style="color:green; font-size:16px">Create new columns **`BONUS`** and **`TOTAL_COMP`**. Use column **`RANDOM_BONUS`** to calculate the bonus.</span>

In [36]:
df_coh['RANDOM_BONUS'] = np.random.rand(len(df_coh)) * .1
df_coh['BONUS'] = df_coh['RANDOM_BONUS']*df_coh['BASE_SALARY']
df_coh['TOTAL_COMP'] = df_coh['RANDOM_BONUS'] + df_coh['BASE_SALARY']

df_coh.iloc[:3, -5:]

EEOJ SNAPSHOT_DATE  RANDOM_BONUS         BONUS  \
0  A: Officials & Administrators    2016-06-01      0.087089  10612.827818   
1          E: Para-Professionals    2016-06-01      0.045993   1201.558852   
2  D: Protective Service Workers    2016-06-01      0.089349   4045.651276   

      TOTAL_COMP  
0  121862.087089  
1   26125.045993  
2   45279.089349

### Problem 18
<span  style="color:green; font-size:16px">Use the **`EXPERIENCE_LEVEL`** column to determine if more experienced employees make more money.</span>

In [44]:
#df_coh['HIRE_DATE'] = df_coh['HIRE_DATE'].astype(np.datetime64)
today = pd.to_datetime('today')
experience = today - df_coh['HIRE_DATE']
years_experience = experience / np.timedelta64(1, 'Y')
df_coh['YEARS_EXPERIENCE'] = years_experience
exp_categories = pd.cut(years_experience, bins=[0, 5, 15, 100], labels=['Novice', 'Experienced', 'Senior'])
df_coh['EXPERIENCE_LEVEL'] = exp_categories

In [ ]:
# your code here
novice = df_coh[df_coh['EXPERIENCE_LEVEL'] == 'Novice']['BASE_SALARY']
exper = df_coh[df_coh['EXPERIENCE_LEVEL'] == 'Experienced']['BASE_SALARY']
senior = df_coh[df_coh['EXPERIENCE_LEVEL'] == 'Senior']['BASE_SALARY']

novice.mean(), exper.mean(), senior.mean()